tests/python_package_test/test_engine.py

```diff
@@ -1545,6 +1545,40 @@ def test_refit():
     assert err_pred > new_err_pred
 
 
+def test_refit_dataset_params():
+    # check refit accepts dataset_params
+    X, y = load_breast_cancer(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    lgb_train = lgb.Dataset(X_train, y_train)
```

Comment:
I've never used `refit` before but it seems that it sets `init_score=0`. So if we use `refit` with the same data and `decay_rate=0.0` I'd expect to get the same results, however that seems to only be true if we set `init_score=0` in the first booster. So with the current test, the assertion about the different scores passes even if we send in the same data to `refit` without modifying any of the dataset arguments.
```suggestion
    lgb_train = lgb.Dataset(X_train, y_train, init_score=np.zeros(y_train.size))
```

Reply:
To illustrate what you say

```python
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import log_loss

import lightgbm as lgb
import numpy as np

X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
#lgb_train = lgb.Dataset(X_train, y_train)
lgb_train = lgb.Dataset(X_train, y_train, init_score=np.zeros(y_train.size))
train_params = {
    'objective': 'binary',
    'verbose': -1,
    'seed': 123
}
gbm = lgb.train(train_params, lgb_train, num_boost_round=10)
non_weight_err_pred = log_loss(y_test, gbm.predict(X_test))
#refit_weight = np.random.rand(y_train.shape[0])
dataset_params = {
    #'max_bin': 260,
    #'min_data_in_bin': 5,
    'data_random_seed': 123,
}
new_gbm  = gbm.refit(
    data=X_train,
    label=y_train,
    #weight=refit_weight,
    dataset_params=dataset_params,
    decay_rate=0.0,
)
weight_err_pred = log_loss(y_test, new_gbm.predict(X_test))
weight_err_pred
>>> 0.2682710111899916
non_weight_err_pred
>>> 0.2682710111899916
```

but if you set train data without init_score `lgb_train = lgb.Dataset(X_train, y_train)` you get different scores

```
weight_err_pred
>>> 0.1442188066230218
non_weight_err_pred
>>> 0.24365628110393137

Reply:
Not sure if this has something to do with how `init_score` is initialized for [_lazy_init](https://github.com/microsoft/LightGBM/blob/44a9120197fbfc7f3bfb9349e4eb2e9443c1971b/python-package/lightgbm/basic.py#L973). Apparently `init_score` will remain `None`.

Reply:
@StrikerRUS @jameslamb any comment on this?, is this related to this pr ?

Reply:
Please wait for @jmoralez to respond to you in this thread. This project is maintained by a team of mostly volunteers who review these PRs on nights and weekends, so please be patient with us.

Reply:
I'd like these changes to be made to the current tests (setting init_score to zero and the decay rate to 0) so we can be sure that the changes in the refitted booster are only due to the different dataset. The fact that refit sets init_score to 0 might be a bug and may be worth opening an issue about it, but for this PR once these changes are made I think it's good to go.

Reply:
ok @jmoralez I've made corresponding changes. For your information, issue #4951 is open if you would like to add more context.